facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+// a whole chunk will be uploaded at once: responses to the PUT requests include
+// a Range header telling you how much of the chunk Google got, so you can build
+// the next PUT request appropriately. So suppose you are trying to upload an 8
+// MiB chunk from the middle of the overall object, and you succeed, but Google
+// tells you they didn't get the last 100 KiB. You might right away want to
+// upload that last 100 KiB, but if you try, you will fail because it's not the
+// final chunk and it's less than 256 KiB. So we do two special things in
+// upload_chunk when we know it's the last chunk: (1) we construct the Content-
+// Range header without any asterisks (2) we drain self.buffer.
+struct StreamingTransferWriter {
+    upload_session_uri: String,
+    minimum_upload_chunk_size: usize,
+    object_upload_position: usize,
+    buffer: Vec<u8>,
+}
+
+impl StreamingTransferWriter {
+    fn new(
+        bucket: String,
+        object: String,
+        minimum_upload_chunk_size: usize,
+        oauth_token: String,
+    ) -> Result<StreamingTransferWriter> {
+        StreamingTransferWriter::new_with_api_url(
+            bucket,
+            object,
+            minimum_upload_chunk_size,
+            oauth_token,
+            STORAGE_API_BASE_URL,
+        )
+    }
+
+    fn new_with_api_url(
+        bucket: String,
+        object: String,
+        minimum_upload_chunk_size: usize,
+        oauth_token: String,
+        storage_api_base_url: &str,
+    ) -> Result<StreamingTransferWriter> {
+        // Initiate the resumable, streaming upload. Unlike the GET side, the
+        // key is provided as a query parameter here, and thus doesn't need to
+        // be URL encoded.
+        // https://cloud.google.com/storage/docs/performing-resumable-uploads#initiate-session
+        let upload_url = format!(
+            "{}/upload/storage/v1/b/{}/o/?uploadType=resumable&name={}",
+            storage_api_base_url, bucket, object
+        );
+        let http_response = ureq::post(&upload_url)
+            .set("Authorization", &format!("Bearer {}", oauth_token))
+            .set("Content-Length", "0")
+            // By default, ureq will wait forever to connect or read
+            .timeout_connect(10_000) // ten seconds
+            .timeout_read(10_000) // ten seconds
+            .call();
+        if http_response.error() {
+            return Err(anyhow!(
+                "failed to initiate streaming transfer: {:?}",
+                http_response
+            ));
+        }
+
+        // The upload session URI authenticates subsequent upload requests for
+        // this upload, so we no longer need the impersonated service account's
+        // Oauth token. Session URIs are valid for a week, which should be more
+        // than enough for any upload we perform.
+        // https://cloud.google.com/storage/docs/resumable-uploads#session-uris
+        let upload_session_uri = match http_response.header("Location") {
```

Comment:
I think this could be:

```rust
let upload_session_uri = http_response.header("Location").unwrap_or_else(|| {
  Err(anyhow!("no Location header in response when initiating streaming transfer"))
})?;
```

Reply:
You're right, but `anyhow::Context` makes this even nicer.